### PR TITLE
Feature/add import controller

### PR DESCRIPTION
**Pull request by x1mx1x**

### CODE REVIEW

File: CharlieBackend.Business/Services/FileImportService.cs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                    };
+
+                    importedGroups.Add(fileLine);
+                    _unitOfWork.StudentGroupRepository.Add(group);
+                    rowCounter++;
+                }
+                catch (FormatException ex)
+                {
+                    _unitOfWork.Rollback();
+                    return Result<List<StudentGroupFileModel>>.Error(ErrorCode.ValidationError,
```
Comment (SerjKocherga) [inline CharlieBackend.Business/Services/FileImportService.cs:83]: Add a new line before return.

File: CharlieBackend.Business/Services/FileImportService.cs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                catch (DbUpdateException ex)
+                {
+                    _unitOfWork.Rollback();
+                    return Result<List<StudentGroupFileModel>>.Error(ErrorCode.ValidationError,
+                        "Inputed data is incorrect.\n" + ex.Message);
+                }
+            }
+
+            await _unitOfWork.CommitAsync();
+            return Result<List<StudentGroupFileModel>>
```
Comment (SerjKocherga) [inline CharlieBackend.Business/Services/FileImportService.cs:95]: Add a new line before return.

File: CharlieBackend.Business/Services/FileImportService.cs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            foreach (Course course in await _unitOfWork.CourseRepository.GetAllAsync())
+            {
+                existingCourseIds.Add(course.Id);
+            }
+
+            foreach (StudentGroup group in await _unitOfWork.StudentGroupRepository.GetAllAsync())
+            {
+                existingGroupNames.Add(group.Name);
+            }
+
```
Comment (SerjKocherga) [inline CharlieBackend.Business/Services/FileImportService.cs:113]: Delete this empty line.

File: CharlieBackend.Business/Services/Interfaces/IFileImportService.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace CharlieBackend.Business.Services.Interfaces
+{
+    public interface IFileImportService
+    {
+        public Task<Result<List<StudentGroupFileModel>>> ImportFileAsync(ImportFileDto file);
+
+        /*
+          
+            protected Task<bool> IsEndOfFileAsync(int rowCounter, IXLWorksheet ws);
```
Comment (SerjKocherga) [inline CharlieBackend.Business/Services/Interfaces/IFileImportService.cs:16]: Why commented code is going to be in dev?

File: CharlieBackend.Business/Services/FileImportService.cs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            if (existingGroupNames.Contains(fileLine.Name))
+            {
+                throw new DbUpdateException($"Group with name {fileLine.Name} already exist.\n" +
+                   $"Problem was occured in col C, row {rowCounter}.");
+            }
+        }
+
+        private async Task<bool> IsEndOfFileAsync(int rowCounter, IXLWorksheet ws)
```
Comment (SerjKocherga) [inline CharlieBackend.Business/Services/FileImportService.cs:146]: Why using async here if you don't have any asynchronous operation inside?

File: CharlieBackend.Api/Controllers/ImportController.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    public class ImportController : ControllerBase
+    {
+        private readonly IFileImportService _importService;
+
+        public ImportController(IFileImportService importService)
+        {
+            _importService = importService;
+        }
+
+
```
Comment (SerjKocherga) [inline CharlieBackend.Api/Controllers/ImportController.cs:22]: Remove this empty new line.

File: CharlieBackend.Api/Controllers/ImportController.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
+﻿
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using CharlieBackend.Core.DTO.File;
+using Microsoft.AspNetCore.Authorization;
+using CharlieBackend.Business.Services.Interfaces;
+using CharlieBackend.Core;
```
Comment (SerjKocherga) [inline CharlieBackend.Api/Controllers/ImportController.cs:7]: Move it to the top.

**Review by SerjKocherga — CHANGES_REQUESTED** [review]

File: CharlieBackend.Business/Services/FileImportService.cs [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+            if (!existingCourseIds.Contains(Convert.ToInt64(fileLine.CourseId)))
+            {
+                throw new DbUpdateException($"Course with id {fileLine.CourseId} doesn't exist.\n" +
+                   $"Problem was occured in col B, row {rowCounter}.");
+            }
+
+            if (existingGroupNames.Contains(fileLine.Name))
+            {
+                throw new DbUpdateException($"Group with name {fileLine.Name} already exist.\n" +
+                   $"Problem was occured in col C, row {rowCounter}.");
```
Comment (SnipeFrost) [inline CharlieBackend.Business/Services/FileImportService.cs:142]: Why are you just throw exceptions instead of using Result<t>?

Comment (SnipeFrost) [reply CharlieBackend.Business/Services/FileImportService.cs:142]: When you throw exception like this you are make fall whole program

Comment (x1mx1x) [reply CharlieBackend.Business/Services/FileImportService.cs:142]: than I'll catch it and returns Results

Comment (x1mx1x) [reply CharlieBackend.Business/Services/FileImportService.cs:142]: this code is in the method that shouldn't returns Results

**Review by Vlad1023 — APPROVED** [review]

**Review by SnipeFrost — APPROVED** [review]

**Review by SerjKocherga — APPROVED** [review]

**Pull request merged**